server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -126,7 +128,17 @@ private Map<String, UploadedSegmentMetadata> readLatestMetadataFile() throws IOE
 
     private Map<String, UploadedSegmentMetadata> readMetadataFile(String metadataFilename) throws IOException {
         try (IndexInput indexInput = remoteMetadataDirectory.openInput(metadataFilename, IOContext.DEFAULT)) {
-            Map<String, String> segmentMetadata = indexInput.readMapOfStrings();
+            // BufferedChecksumIndexInput reader keeps computing checksum for all the bytes we read from indexInput
+            // This checksum is then compared with checksum in footer
+            ChecksumIndexInput in = new BufferedChecksumIndexInput(indexInput);
+            CodecUtil.checkHeader(
+                in,
+                UploadedSegmentMetadata.METADATA_CODEC,
+                UploadedSegmentMetadata.CURRENT_VERSION,
+                UploadedSegmentMetadata.CURRENT_VERSION
```

Comment:
we could try keeping a mapping of OS version to this version like this - https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/Version.java#L79-L93

Reply:
Thanks @ashking94. Let me check on this more and see if we can follow a similar approach with metadata versions. We would definitely have to maintain some version mapping b/w metadata versions and OS versions. But i was thinking about how can we provide maximum compatibility across versions and better backward compatibility for data migration. I know lucene already adds some restriction to that, but for metadata files maybe we can be a little flexible.